model/scratch/dffml_model_scratch/model/slr.py

```diff
@@ -0,0 +1,120 @@
+        squared_error_regression = await self.squared_error(ys, regression_line)
+        return 1 - (squared_error_regression/squared_error_mean)
+
+    async def applicable_features(self, features: Features):
+        usable = []
+        if len(features) != 1:
+            raise ValueError("Simple Linear Regression doesn't support features other than 1")
+        for feature in features:
+            if feature.dtype() != int and feature.dtype() != float:
+                raise ValueError("Simple Linear Regression only supports int or float feature")
+            if feature.length() != 1:
+                raise ValueError("Simple LR only supports single values (non-matrix / array)")
+            usable.append(feature.NAME)
+        return usable
+
+    async def best_fit_line(self, xs, ys):
+        m = ((mean(xs)*mean(ys)) - mean(xs*ys))/((mean(xs)*mean(xs)) - (mean(xs*xs)))
+        b = mean(ys) - (m*mean(xs))
+        regression_line = [m*x + b for x in xs]
+        accuracy = await self.coeff_of_deter(ys, regression_line)
+        return (m, b, accuracy)
+
+    async def predict_input(self, x):
+        prediction = self.regression_line[0]*x+self.regression_line[1]
+        return prediction
+
+    async def train(self, sources: Sources, features: Features):
+        '''
+        Train using repos as the data to learn from.
+        '''
+        feature = await self.applicable_features(features)
+        async for repo in sources.with_features(feature):
+            # Grab a subset of the feature data being stored within the repo
+            # The subset is the feature_we_care_about and the feature we are want to predict
+            feature_data = repo.features(feature + [self.parent.config.predict])
+            self.xData = np.append(self.xData, feature_data[feature[0]])
+            self.yData = np.append(self.yData, feature_data[self.parent.config.predict])
+        self.regression_line = await self.best_fit_line(self.xData, self.yData)
+        
+    async def accuracy(self, sources: Sources, features: Features) -> Accuracy:
+        '''
+        Evaluates the accuracy of our model after training using the input repos
+        as test data.
+        '''
+        if self.regression_line is None:
+            raise ValueError('Model Not Trained')
+        accuracy_value = self.regression_line[2]
+        return Accuracy(accuracy_value)
+
+    async def predict(self, repos: AsyncIterator[Repo], features: Features) -> \
+                    AsyncIterator[Tuple[Repo, Any, float]]:
+        '''
+        Uses trained data to make a prediction about the quality of a repo.
+        '''
+        feature = await self.applicable_features(features)
+        async for repo in repos:
+            feature_data = repo.features(feature)
+            yield repo, self.predict_input(feature_data[feature[0]]), self.regression_line[2]
+
+
+@entry_point('slr')
+class SLR(Model):
+
+    CONTEXT = SLRContext
```

Comment:
For saving an loading from disk, you'll need to implement these methods

```python
class SLR(Model):

    def _filename(self):
        # Create a file path that is specific to the data we are predicting on
        return os.path.join(self.parent.config.directory, hashlib.sha384(self.config.predict.encode()).hexdigest() + '.json')

    async def __aenter__(self) -> "SLRContext":
        filename = self._filename()
        if os.path.isfile(filename):
            # Load from file using open, and json.load. You'll want to make a self.saved property (in __init__ or something)
        return self

    async def __aexit__(self, exc_type, exc_value, traceback):
        # Save to the file using open, and json.dump. Dump out the self.saved property to the file.
        pass
```

Then in the context, you will want to access `self.parent.saved`, to save and load your regression line. Probably something like this (turn it into a method):

```python
class SLRContext(ModelContext):

    def get_regression_line(self, features: List[str]):
        # Try to get the line, if it doesn't exist in the saved structure return None
        return self.parent.saved.get(hashlib.sha384((''.join(sorted(features))).encode()).hexdigest(), None)

    def set_regression_line(self, features: List[str], slope, constant, accuracy):
        return self.parent.saved[hashlib.sha384((''.join(sorted(features))).encode()).hexdigest()] = (slope, constant, accuracy)
```

> There should probably be a helper method for `hashlib.sha384((''.join(sorted(features))).encode()).hexdigest()` too so if it's changed one place we don't forget to change it the other place

Reply:
On It!!!